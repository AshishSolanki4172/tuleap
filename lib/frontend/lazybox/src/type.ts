/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { HTMLTemplateResult, TemplateResult } from "lit/html.js";
import type { GroupCollection, LazyboxItem } from "./items/GroupCollection";
import type { HTMLTemplateStringProcessor } from "./index";

export interface Lazybox {
    setDropdownContent: (groups: GroupCollection) => void;
    resetSelection: () => void;
    setSelection: (selection: ReadonlyArray<LazyboxItem>) => void;
    destroy: () => void;
}

export type LazyboxSearchFieldCallback = (query: string) => void;

export type LazyboxSelectionCallback = (selected_value: unknown | null) => void;

export type LazyboxTemplatingCallback = (
    html: typeof HTMLTemplateStringProcessor,
    item: LazyboxItem
) => HTMLTemplateResult;

export type LazyboxNewItemCallback = () => void;

export type LazyBoxWithNewItemButton = {
    readonly new_item_button_label: string;
    readonly new_item_callback: LazyboxNewItemCallback;
};

type LazyBoxWithoutNewItemButton = {
    readonly new_item_callback?: undefined;
};

export type LazyboxOptions = (LazyBoxWithoutNewItemButton | LazyBoxWithNewItemButton) & {
    readonly placeholder: string;
    readonly search_input_placeholder: string;
    readonly is_multiple: boolean;
    readonly templating_callback: LazyboxTemplatingCallback;
    readonly selection_callback: LazyboxSelectionCallback;
    readonly search_field_callback: LazyboxSearchFieldCallback;
};

export type RenderedItemMap = Map<string, RenderedItem>;

export interface RenderedItem {
    id: string;
    template: TemplateResult;
    value: unknown;
    is_disabled: boolean;
    is_selected: boolean;
    group_id: string;
    element: Element;
}

export interface LazyboxComponent {
    wrapper_element: HTMLElement;
    lazybox_element: Element;
    dropdown_element: HTMLElement;
    selection_element: HTMLElement;
    placeholder_element: Element;
    dropdown_list_element: HTMLElement;
    search_field_element: HTMLInputElement;
}

export interface LazyboxSelectionStateSingle {
    selected_item: RenderedItem;
    selected_value_element: DocumentFragment;
}

export interface LazyboxSelectionStateMultiple {
    selected_items: Map<string, RenderedItem>;
    selected_values_elements: Map<string, Element>;
}

export interface ScrollCoordinates {
    x_position: number;
    y_position: number;
}

export interface ManageSelection {
    processSelection(item: Element): void;
    hasSelection(): boolean;
    updateSelectionAfterDropdownContentChange(): void;
    clearSelection(): void;
    setSelection(selection: ReadonlyArray<RenderedItem>): void;
}
