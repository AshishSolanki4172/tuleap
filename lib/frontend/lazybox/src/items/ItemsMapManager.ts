/*
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { RenderedItem, RenderedItemMap } from "../type";
import type { ListItemMapBuilderType } from "./ListItemMapBuilder";
import type { GroupCollection } from "./GroupCollection";

export class ItemsMapManager {
    private items_map!: RenderedItemMap;

    constructor(private readonly list_item_builder: ListItemMapBuilderType) {}

    public getLazyboxItems(): ReadonlyArray<RenderedItem> {
        return [...this.items_map.values()];
    }

    public findLazyboxItemInItemMap(item_id: string): RenderedItem {
        const list_item = this.items_map.get(item_id);
        if (!list_item) {
            throw new Error(`Item with id ${item_id} not found in item map`);
        }
        return list_item;
    }

    public refreshItemsMap(groups: GroupCollection): void {
        this.items_map = this.list_item_builder.buildLazyboxItemsMap(groups);
    }

    public getItemWithValue(value: unknown): RenderedItem | null {
        return this.getLazyboxItems().find((item) => item.value === value) ?? null;
    }
}
