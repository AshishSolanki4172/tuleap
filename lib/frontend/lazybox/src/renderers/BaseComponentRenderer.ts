/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */
import type { LazyboxComponent } from "../type";
import { isLazyboxInAModal } from "../helpers/lazybox-in-modals-helper";

export class BaseComponentRenderer {
    constructor(
        private readonly doc: Document,
        private readonly source_select_box: HTMLSelectElement,
        private readonly placeholder: string,
        private readonly input_placeholder: string,
        private readonly is_multiple: boolean
    ) {}

    public renderBaseComponent(): LazyboxComponent {
        const wrapper_element = document.createElement("span");
        wrapper_element.classList.add("lazybox-component-wrapper");

        const lazybox_element = this.createLazyboxElement();
        const dropdown_element = this.createDropdownElement();
        const dropdown_list_element = this.createDropdownListElement();
        const selection_element = this.createSelectionElement();
        const placeholder_element = this.createPlaceholderElement();
        const search_field_element = this.createSearchFieldElement();

        if (!this.source_select_box.disabled) {
            selection_element.setAttribute("tabindex", "0");
        }

        if (this.is_multiple) {
            search_field_element.setAttribute("placeholder", this.placeholder);
            const new_search_section = this.createSearchSectionForMultipleListPicker();
            new_search_section.appendChild(search_field_element);
            selection_element.appendChild(new_search_section);
        } else {
            const search_section_element = this.createSearchSectionElement();
            search_section_element.appendChild(search_field_element);
            dropdown_element.appendChild(search_section_element);
            selection_element.appendChild(placeholder_element);
        }

        lazybox_element.appendChild(selection_element);
        dropdown_element.appendChild(dropdown_list_element);
        wrapper_element.appendChild(lazybox_element);
        this.doc.body.insertAdjacentElement("beforeend", dropdown_element);

        this.source_select_box.insertAdjacentElement("afterend", wrapper_element);

        if (isLazyboxInAModal(wrapper_element)) {
            dropdown_element.classList.add("lazybox-dropdown-over-modal");
        }

        return {
            wrapper_element,
            lazybox_element: lazybox_element,
            dropdown_element,
            selection_element,
            placeholder_element,
            dropdown_list_element,
            search_field_element,
        };
    }

    private createDropdownListElement(): HTMLElement {
        const dropdown_list_element = document.createElement("ul");
        dropdown_list_element.classList.add("lazybox-dropdown-values-list");
        dropdown_list_element.setAttribute("role", "listbox");
        dropdown_list_element.setAttribute("aria-expanded", "false");
        dropdown_list_element.setAttribute("aria-hidden", "false");
        return dropdown_list_element;
    }

    private createDropdownElement(): HTMLElement {
        const dropdown_element = document.createElement("span");
        dropdown_element.classList.add("lazybox-dropdown");
        dropdown_element.setAttribute("data-test", "lazybox-dropdown");
        return dropdown_element;
    }

    private createPlaceholderElement(): Element {
        const placeholder_element = document.createElement("span");
        placeholder_element.classList.add("lazybox-placeholder");
        placeholder_element.appendChild(document.createTextNode(this.placeholder));
        return placeholder_element;
    }

    private createSelectionElement(): HTMLElement {
        const selection_element = document.createElement("span");
        selection_element.classList.add("lazybox-selection");
        selection_element.setAttribute("data-test", "lazybox-selection");

        if (this.is_multiple) {
            selection_element.classList.add("lazybox-multiple");
            selection_element.setAttribute("aria-haspopup", "true");
            selection_element.setAttribute("aria-expanded", "false");
            selection_element.setAttribute("role", "combobox");
        } else {
            selection_element.classList.add("lazybox-single");
            selection_element.setAttribute("role", "textbox");
            selection_element.setAttribute("aria-readonly", "true");
        }

        return selection_element;
    }

    private createLazyboxElement(): Element {
        const lazybox_element = document.createElement("span");
        lazybox_element.classList.add("lazybox");
        lazybox_element.setAttribute("data-test", "lazybox");

        if (this.source_select_box.disabled) {
            lazybox_element.classList.add("lazybox-disabled");
        }

        return lazybox_element;
    }

    private createSearchSectionElement(): Element {
        const search_section = document.createElement("span");
        search_section.classList.add("lazybox-single-dropdown-search-section");

        return search_section;
    }

    private createSearchSectionForMultipleListPicker(): Element {
        const search_section = document.createElement("span");

        search_section.classList.add("lazybox-multiple-search-section");
        return search_section;
    }

    private createSearchFieldElement(): HTMLInputElement {
        const search_field_element = document.createElement("input");
        if (this.source_select_box.disabled) {
            search_field_element.setAttribute("disabled", "disabled");
        }
        search_field_element.setAttribute("data-test", "lazybox-search-field");
        search_field_element.classList.add("lazybox-search-field");
        search_field_element.setAttribute("type", "search");
        search_field_element.setAttribute("tabindex", "0");
        search_field_element.setAttribute("autocomplete", "off");
        search_field_element.setAttribute("autocorrect", "off");
        search_field_element.setAttribute("autocapitalize", "none");
        search_field_element.setAttribute("spellcheck", "none");
        search_field_element.setAttribute("role", "searchbox");
        search_field_element.setAttribute("aria-autocomplete", "list");
        search_field_element.setAttribute("aria-controls", "lazybox-dropdown-values-list");
        search_field_element.setAttribute("placeholder", this.input_placeholder);

        return search_field_element;
    }
}
