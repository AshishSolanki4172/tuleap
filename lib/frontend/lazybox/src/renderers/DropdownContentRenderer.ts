/*
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import { html, render } from "lit/html.js";
import type { HTMLTemplateResult } from "lit/html.js";
import type { ItemsMapManager } from "../items/ItemsMapManager";
import type { GroupCollection } from "../items/GroupCollection";
import { getGroupId } from "../helpers/group-id-helper";
import type { GroupOfItems } from "../items/GroupCollection";

export const getRenderedListItem = (
    option_id: string,
    template: HTMLTemplateResult,
    is_disabled: boolean
): Element => {
    let class_name = "lazybox-dropdown-option-value";
    if (is_disabled) {
        class_name = "lazybox-dropdown-option-value-disabled";
    }

    const document_fragment = document.createDocumentFragment();
    render(
        html`
            <li
                role="option"
                aria-selected="false"
                data-item-id="${option_id}"
                class="${class_name}"
                data-test="lazybox-item"
            >
                ${template}
            </li>
        `,
        document_fragment
    );

    const list_item = document_fragment.firstElementChild;
    if (list_item !== null) {
        return list_item;
    }

    throw new Error("Cannot render the list item");
};

const createEmptyDropdownState = (dropdown_message: string): HTMLTemplateResult =>
    html`
        <li
            class="lazybox-empty-dropdown-state"
            role="alert"
            aria-live="assertive"
            data-test="lazybox-empty-state"
        >
            ${dropdown_message}
        </li>
    `;

const getGroupLabel = (group: GroupOfItems): HTMLTemplateResult => {
    if (group.is_loading) {
        return html`
            <strong class="lazybox-group-label">
                ${group.label}
                <i
                    class="fa-solid fa-spin fa-circle-notch lazybox-loading-group-spinner"
                    data-test="lazybox-loading-group-spinner"
                ></i>
            </strong>
        `;
    }

    return html`<strong class="lazybox-group-label">${group.label}</strong>`;
};

const getGroupFooter = (group: GroupOfItems): HTMLTemplateResult =>
    group.footer_message !== "" && group.items.length > 0
        ? html`<p class="lazybox-group-footer" data-test="lazybox-group-footer">
              ${group.footer_message}
          </p>`
        : html``;

export class DropdownContentRenderer {
    constructor(
        private readonly dropdown_list_element: HTMLElement,
        private readonly items_map_manager: ItemsMapManager
    ) {}

    public renderLazyboxDropdownContent(groups: GroupCollection): void {
        const group_templates = groups.map((group) => {
            const group_id = getGroupId(group);
            const lazybox_items = this.items_map_manager
                .getLazyboxItems()
                .filter((item) => item.group_id === group_id);

            const items_template =
                group.items.length === 0
                    ? createEmptyDropdownState(group.empty_message)
                    : lazybox_items.map((item) => item.element);

            return html`
                <li class="lazybox-item-group">
                    ${getGroupLabel(group)}
                    <ul
                        role="group"
                        aria-label="${group.label}"
                        class="lazybox-dropdown-option-group"
                    >
                        ${items_template}
                    </ul>
                    ${getGroupFooter(group)}
                </li>
            `;
        });

        render(group_templates, this.dropdown_list_element);
    }
}
