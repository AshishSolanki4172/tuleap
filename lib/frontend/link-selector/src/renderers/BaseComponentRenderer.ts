/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */
import type { LinkSelectorComponent } from "../type";
import { isLinkSelectorInAModal } from "../helpers/link-selector-in-modals-helper";

export class BaseComponentRenderer {
    constructor(
        private readonly doc: Document,
        private readonly source_select_box: HTMLSelectElement,
        private readonly placeholder: string,
        private readonly input_placeholder: string,
        private readonly is_multiple: boolean
    ) {}

    public renderBaseComponent(): LinkSelectorComponent {
        const wrapper_element = document.createElement("span");
        wrapper_element.classList.add("link-selector-component-wrapper");

        const link_selector_element = this.createLinkSelectorElement();
        const dropdown_element = this.createDropdownElement();
        const dropdown_list_element = this.createDropdownListElement();
        const selection_element = this.createSelectionElement();
        const placeholder_element = this.createPlaceholderElement();
        const search_field_element = this.createSearchFieldElement();

        if (!this.source_select_box.disabled) {
            selection_element.setAttribute("tabindex", "0");
        }

        if (this.is_multiple) {
            search_field_element.setAttribute("placeholder", this.placeholder);
            const new_search_section = this.createSearchSectionForMultipleListPicker();
            new_search_section.appendChild(search_field_element);
            selection_element.appendChild(new_search_section);
        } else {
            const search_section_element = this.createSearchSectionElement();
            search_section_element.appendChild(search_field_element);
            dropdown_element.appendChild(search_section_element);
            selection_element.appendChild(placeholder_element);
        }

        link_selector_element.appendChild(selection_element);
        dropdown_element.appendChild(dropdown_list_element);
        wrapper_element.appendChild(link_selector_element);
        this.doc.body.insertAdjacentElement("beforeend", dropdown_element);

        this.source_select_box.insertAdjacentElement("afterend", wrapper_element);

        if (isLinkSelectorInAModal(wrapper_element)) {
            dropdown_element.classList.add("link-selector-dropdown-over-modal");
        }

        return {
            wrapper_element,
            link_selector_element,
            dropdown_element,
            selection_element,
            placeholder_element,
            dropdown_list_element,
            search_field_element,
        };
    }

    private createDropdownListElement(): HTMLElement {
        const dropdown_list_element = document.createElement("ul");
        dropdown_list_element.classList.add("link-selector-dropdown-values-list");
        dropdown_list_element.setAttribute("role", "listbox");
        dropdown_list_element.setAttribute("aria-expanded", "false");
        dropdown_list_element.setAttribute("aria-hidden", "false");
        return dropdown_list_element;
    }

    private createDropdownElement(): HTMLElement {
        const dropdown_element = document.createElement("span");
        dropdown_element.classList.add("link-selector-dropdown");
        dropdown_element.setAttribute("data-test", "link-selector-dropdown");
        return dropdown_element;
    }

    private createPlaceholderElement(): Element {
        const placeholder_element = document.createElement("span");
        placeholder_element.classList.add("link-selector-placeholder");
        placeholder_element.appendChild(document.createTextNode(this.placeholder));
        return placeholder_element;
    }

    private createSelectionElement(): HTMLElement {
        const selection_element = document.createElement("span");
        selection_element.classList.add("link-selector-selection");
        selection_element.setAttribute("data-test", "link-selector-selection");

        if (this.is_multiple) {
            selection_element.classList.add("link-selector-multiple");
            selection_element.setAttribute("aria-haspopup", "true");
            selection_element.setAttribute("aria-expanded", "false");
            selection_element.setAttribute("role", "combobox");
        } else {
            selection_element.classList.add("link-selector-single");
            selection_element.setAttribute("role", "textbox");
            selection_element.setAttribute("aria-readonly", "true");
        }

        return selection_element;
    }

    private createLinkSelectorElement(): Element {
        const link_selector_element = document.createElement("span");
        link_selector_element.classList.add("link-selector");
        link_selector_element.setAttribute("data-test", "link-selector");

        if (this.source_select_box.disabled) {
            link_selector_element.classList.add("link-selector-disabled");
        }

        return link_selector_element;
    }

    private createSearchSectionElement(): Element {
        const search_section = document.createElement("span");
        search_section.classList.add("link-selector-single-dropdown-search-section");

        return search_section;
    }

    private createSearchSectionForMultipleListPicker(): Element {
        const search_section = document.createElement("span");

        search_section.classList.add("link-selector-multiple-search-section");
        return search_section;
    }

    private createSearchFieldElement(): HTMLInputElement {
        const search_field_element = document.createElement("input");
        if (this.source_select_box.disabled) {
            search_field_element.setAttribute("disabled", "disabled");
        }
        search_field_element.setAttribute("data-test", "link-selector-search-field");
        search_field_element.classList.add("link-selector-search-field");
        search_field_element.setAttribute("type", "search");
        search_field_element.setAttribute("tabindex", "0");
        search_field_element.setAttribute("autocomplete", "off");
        search_field_element.setAttribute("autocorrect", "off");
        search_field_element.setAttribute("autocapitalize", "none");
        search_field_element.setAttribute("spellcheck", "none");
        search_field_element.setAttribute("role", "searchbox");
        search_field_element.setAttribute("aria-autocomplete", "list");
        search_field_element.setAttribute("aria-controls", "link-selector-dropdown-values-list");
        search_field_element.setAttribute("placeholder", this.input_placeholder);

        return search_field_element;
    }
}
