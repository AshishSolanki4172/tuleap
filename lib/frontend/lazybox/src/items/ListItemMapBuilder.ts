/*
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { LazyboxTemplatingCallback, RenderedItem, RenderedItemMap } from "../type";
import { getRenderedListItem } from "../renderers/DropdownContentRenderer";
import type { GroupCollection, LazyboxItem } from "./GroupCollection";
import { getGroupId } from "../helpers/group-id-helper";
import { html } from "lit/html.js";

const getItemId = (item: LazyboxItem, group_id: string): string => {
    let base_id = "lazybox-item-";

    if (group_id !== "") {
        base_id += group_id + "-";
    }

    return base_id + item.id;
};

export interface ListItemMapBuilderType {
    buildLazyboxItemsMap(groups: GroupCollection): RenderedItemMap;
    buildRenderedItem(item: LazyboxItem, group_id: string): RenderedItem;
}

export const ListItemMapBuilder = (
    templating_callback: LazyboxTemplatingCallback
): ListItemMapBuilderType => {
    const buildRenderedItem = (item: LazyboxItem, group_id: string): RenderedItem => {
        const id = getItemId(item, group_id);
        const template = templating_callback(html, item);
        return {
            id,
            group_id,
            is_disabled: item.is_disabled,
            is_selected: false,
            value: item.value,
            template: template,
            element: getRenderedListItem(id, template, item.is_disabled),
        };
    };

    const addItemInMap = (
        group_id: string,
        item: LazyboxItem,
        accumulator: RenderedItemMap
    ): RenderedItemMap => {
        const rendered_item = buildRenderedItem(item, group_id);
        accumulator.set(rendered_item.id, rendered_item);
        return accumulator;
    };

    return {
        buildLazyboxItemsMap(groups: GroupCollection): RenderedItemMap {
            return groups.reduce((accumulator: RenderedItemMap, group) => {
                const group_id = getGroupId(group);
                return group.items.reduce(
                    (inner_accumulator: RenderedItemMap, item) =>
                        addItemInMap(group_id, item, inner_accumulator),
                    accumulator
                );
            }, new Map());
        },
        buildRenderedItem,
    };
};
