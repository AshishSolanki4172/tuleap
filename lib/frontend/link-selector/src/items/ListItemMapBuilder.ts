/*
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { RenderedItem, RenderedItemMap } from "../type";
import { getRenderedListItem } from "../renderers/DropdownContentRenderer";
import type { GroupCollection, LinkSelectorItem } from "./GroupCollection";
import { getGroupId } from "../helpers/group-id-helper";

const getItemId = (value: string, group_id: string): string => {
    let base_id = "link-selector-item-";

    if (group_id !== "") {
        base_id += group_id + "-";
    }

    const option_value = value.includes(" ") ? value.split(" ").join("-") : value;

    return base_id + option_value;
};

const buildOption = (doc: Document, item: LinkSelectorItem, id: string): HTMLOptionElement => {
    const option = doc.createElement("option");
    option.value = item.value;
    option.setAttribute("data-item-id", id);
    return option;
};

const addItemInMap = (
    doc: Document,
    group_id: string,
    item: LinkSelectorItem,
    accumulator: RenderedItemMap
): RenderedItemMap => {
    const id = getItemId(item.value, group_id);
    const option = buildOption(doc, item, id);
    const link_selector_item: RenderedItem = {
        id,
        group_id,
        is_disabled: false,
        is_selected: false,
        value: item.value,
        target_option: option,
        template: item.template,
        element: getRenderedListItem(id, item.template, false),
    };
    accumulator.set(id, link_selector_item);
    return accumulator;
};

export class ListItemMapBuilder {
    public buildLinkSelectorItemsMap(groups: GroupCollection): RenderedItemMap {
        return groups.reduce((accumulator: RenderedItemMap, group) => {
            const group_id = getGroupId(group);
            return group.items.reduce(
                (inner_accumulator: RenderedItemMap, item) =>
                    addItemInMap(document, group_id, item, inner_accumulator),
                accumulator
            );
        }, new Map());
    }
}
