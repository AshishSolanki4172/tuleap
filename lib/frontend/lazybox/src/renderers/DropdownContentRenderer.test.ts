/*
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import { describe, it, expect, beforeEach } from "vitest";
import { html } from "lit/html.js";
import { DropdownContentRenderer, getRenderedListItem } from "./DropdownContentRenderer";
import { BaseComponentRenderer } from "./BaseComponentRenderer";
import { ItemsMapManager } from "../items/ItemsMapManager";
import { ListItemMapBuilder } from "../items/ListItemMapBuilder";
import { GroupCollectionBuilder } from "../../tests/builders/GroupCollectionBuilder";
import type { GroupCollection } from "../items/GroupCollection";
import { TemplatingCallbackStub } from "../../tests/stubs/TemplatingCallbackStub";
import { OptionsBuilder } from "../../tests/builders/OptionsBuilder";

describe("DropDownContentRenderer", () => {
    let dropdown: Element, dropdown_list: HTMLElement;

    const render = (groups: GroupCollection): void => {
        const items_map_manager = new ItemsMapManager(
            ListItemMapBuilder(TemplatingCallbackStub.build())
        );
        items_map_manager.refreshItemsMap(groups);
        const renderer = new DropdownContentRenderer(dropdown_list, items_map_manager);
        return renderer.renderLazyboxDropdownContent(groups);
    };

    describe("rendering", () => {
        beforeEach(() => {
            const doc = document.implementation.createHTMLDocument();
            const select = doc.createElement("select");
            const { dropdown_element, dropdown_list_element } = new BaseComponentRenderer(
                document.implementation.createHTMLDocument(),
                select,
                OptionsBuilder.withoutNewItem().build()
            ).renderBaseComponent();

            dropdown = dropdown_element;
            dropdown_list = dropdown_list_element;
        });

        it("renders grouped list items", () => {
            render(GroupCollectionBuilder.withTwoGroups());
            expect(stripExpressionComments(dropdown.innerHTML)).toMatchSnapshot();
        });

        it("renders empty option groups when they have a placeholder", () => {
            const empty_state_text = "No results found on the server";
            const groups = GroupCollectionBuilder.withSingleGroup({
                empty_message: empty_state_text,
                items: [],
            });

            render(groups);

            const empty_state = dropdown_list.querySelector("[data-test=lazybox-empty-state]");
            if (!empty_state) {
                throw new Error("The empty state has not been found in the dropdown");
            }
            expect(empty_state.textContent?.trim()).toBe(empty_state_text);
        });

        it(`renders group footer message below the list of items`, () => {
            const footer_message = "Maybe there are more results";
            const groups = GroupCollectionBuilder.withSingleGroup({
                items: [
                    { id: "value-1", value: { id: 1 }, is_disabled: false },
                    { id: "value-2", value: { id: 2 }, is_disabled: false },
                ],
                footer_message,
            });

            render(groups);

            const footer = dropdown_list.querySelector("[data-test=lazybox-group-footer]");
            expect(footer?.textContent?.trim()).toBe(footer_message);
        });

        it("renders a spinner next to the group title when it is loading", () => {
            const empty_state_text = "I am loading, wait a second!";
            const groups = GroupCollectionBuilder.withSingleGroup({
                label: "A group still loading",
                empty_message: empty_state_text,
                items: [],
                is_loading: true,
            });

            render(groups);

            const spinner = dropdown_list.querySelector(
                "[data-test=lazybox-loading-group-spinner]"
            );
            expect(spinner).toBeDefined();
        });
    });

    describe(`getRenderedListItem()`, () => {
        it(`renders a list item from a given template`, () => {
            const item = getRenderedListItem(
                "value_1",
                html`
                    <span>Badge</span>
                    Value 1
                `,
                false
            );

            expect(stripExpressionComments(item.outerHTML)).toMatchInlineSnapshot(`
                <li role="option" aria-selected="false" data-test="lazybox-item" data-item-id="value_1" class="lazybox-dropdown-option-value">

                  <span>Badge</span>
                  Value 1

                </li>
            `);
        });

        it(`renders a disabled list item`, () => {
            const item = getRenderedListItem(
                "value_1",
                html`
                    <span>Badge</span>
                    Value 1
                `,
                true
            );

            expect(item.classList.contains("lazybox-dropdown-option-value-disabled")).toBe(true);
        });
    });
});

/**
 * See https://github.com/lit/lit/blob/lit%402.0.2/packages/lit-html/src/test/test-utils/strip-markers.ts
 */
function stripExpressionComments(html: string): string {
    return html.replace(/<!--\?lit\$[0-9]+\$-->|<!--\??-->/g, "");
}
