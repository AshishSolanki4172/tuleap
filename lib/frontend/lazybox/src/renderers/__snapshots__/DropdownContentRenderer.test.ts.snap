// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DropDownContentRenderer > rendering > renders grouped list items 1`] = `
<span class="lazybox-single-dropdown-search-section"><input data-test="lazybox-search-field" class="lazybox-search-field" type="search" tabindex="0" autocomplete="off" autocorrect="off" autocapitalize="none" spellcheck="none" role="searchbox" aria-autocomplete="list" aria-controls="lazybox-dropdown-values-list" placeholder=""></span>
<ul class="lazybox-dropdown-values-list" role="listbox" aria-expanded="false" aria-hidden="false">
  <li class="lazybox-item-group">
    <strong class="lazybox-group-label">Group 1</strong>
    <ul role="group" class="lazybox-dropdown-option-group" aria-label="Group 1">
      <li role="option" aria-selected="false" data-test="lazybox-item" data-item-id="lazybox-item-group1-value-0" class="lazybox-dropdown-option-value">
        Value 0
      </li>
      <li role="option" aria-selected="false" data-test="lazybox-item" data-item-id="lazybox-item-group1-value-1" class="lazybox-dropdown-option-value">
        Value 1
      </li>
      <li role="option" aria-selected="false" data-test="lazybox-item" data-item-id="lazybox-item-group1-value-2" class="lazybox-dropdown-option-value">
        Value 2
      </li>
    </ul>

  </li>

  <li class="lazybox-item-group">
    <strong class="lazybox-group-label">Group 2</strong>
    <ul role="group" class="lazybox-dropdown-option-group" aria-label="Group 2">
      <li role="option" aria-selected="false" data-test="lazybox-item" data-item-id="lazybox-item-group2-value-3" class="lazybox-dropdown-option-value">
        Value 3
      </li>
      <li role="option" aria-selected="false" data-test="lazybox-item" data-item-id="lazybox-item-group2-value-4" class="lazybox-dropdown-option-value">
        Value 4
      </li>
      <li role="option" aria-selected="false" data-test="lazybox-item" data-item-id="lazybox-item-group2-value-5" class="lazybox-dropdown-option-value-disabled">
        Value 5
      </li>
    </ul>

  </li>
</ul>
`;
